src/Microsoft.TestPlatform.Common/Filtering/FilterExpression.cs

```diff
@@ -127,30 +127,59 @@ private static void ProcessOperator(Stack<FilterExpression> filterStack, Operato
         }
 
-        bool valid = false;
-        if (_condition != null)
+        FilterExpression? current = this;
+        Stack<FilterExpression> filterStack = new();
+        Stack<string[]?> result = new();
+        do
         {
-            valid = _condition.ValidForProperties(properties, propertyProvider);
-            if (!valid)
+            //Push root's right child and then root to stack then Set root as root's left child
+            while (current != null)
             {
-                invalidProperties = new string[1] { _condition.Name };
+                if (current._right != null)
+                {
+                    filterStack.Push(current._right);
+                }
+                filterStack.Push(current);
+                current = current._left;
             }
-        }
-        else
-        {
-            invalidProperties = _left!.ValidForProperties(properties, propertyProvider);
-            var invalidRight = _right!.ValidForProperties(properties, propertyProvider);
-            if (null == invalidProperties)
+            //If the popped item has a right child and the right child is at top of stack, then remove the right child from stack, push the root back and set root as root's right child.
+            current = filterStack.Pop();
+            if (filterStack.Count > 0 && current._right == filterStack.First())
             {
-                invalidProperties = invalidRight;
+                filterStack.Pop();
+                filterStack.Push(current);
+                current = current._right;
             }
-            else if (null != invalidRight)
+            else
             {
-                invalidProperties = invalidProperties.Concat(invalidRight).ToArray();
+                bool valid = false;
+                if (current._condition != null)
+                {
+                    valid = current._condition.ValidForProperties(properties, propertyProvider);
+                    if (!valid)
+                    {
+                        invalidProperties = new string[1] { current._condition.Name };
+                    }
+                }
+                else
+                {
+                    var invalidRight = current._right != null ? result.Pop() : null;
+                    invalidProperties = current._left != null ? result.Pop() : null;
+                    if (null == invalidProperties)
+                    {
+                        invalidProperties = invalidRight;
+                    }
+                    else if (null != invalidRight)
+                    {
+                        invalidProperties = invalidProperties.Concat(invalidRight).ToArray();
+                    }
+                }
+                result.Push(invalidProperties);
+                current = null;
             }
-        }
+        } while (filterStack.Count() > 0);
 
-        return invalidProperties;
+        return result.First(); // the result stack will have one element at the end
```

Comment:
Could we add an assertion like `TPDebug.Assert(result.Count == 1, "Result stack should have one element at the end.");`. Not sure if we are supposed to have 1+ or only 1.